File: src/main/java/dev/lbuddyboy/flash/user/impl/RedisUser.java
package dev.lbuddyboy.flash.user.impl;

import dev.lbuddyboy.flash.handler.RedisHandler;
import dev.lbuddyboy.flash.user.User;
import dev.lbuddyboy.flash.user.model.Grant;
import dev.lbuddyboy.flash.user.model.UserPermission;
import dev.lbuddyboy.flash.util.Tasks;
import dev.lbuddyboy.flash.util.gson.GSONUtils;
import lombok.Data;
import lombok.RequiredArgsConstructor;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class RedisUser extends User {

    public RedisUser(UUID uuid, String name, boolean load) {
        this.uuid = uuid;
        this.name = name;

        if (load) load();
    }

    @Override
    public void load() {
        try {
            RedisUser user = GSONUtils.getGSON().fromJson(RedisHandler.requestJedis().getResource().hget("Users", getUuid().toString()), GSONUtils.REDIS_USER);

            this.name = user.getName();
            this.ip = user.getIp();
            this.lastServer = user.getLastServer();
            this.permissions = user.getPermissions();
            this.knownIps = user.getKnownIps();
            this.grants = user.getGrants();

            if (this.grants.isEmpty()) {
                this.grants.add(Grant.defaultGrant());
            }
        } catch (Exception ignored) {
            save(true, true);
        }
    }

    @Override
    public void save(boolean async) {
        save(async, false);
    }

    private void save(boolean async, boolean reload) {
        if (!async) {
            RedisHandler.requestJedis().getResource().hset("Users", getUuid().toString(), toJson());
            if (reload) load();
            return;
        }
        Tasks.runAsync(() -> save(false, reload));
    }

    public String toJson() {
        return GSONUtils.getGSON().toJson(this, GSONUtils.REDIS_USER);
    }

}

File: src/main/java/dev/lbuddyboy/flash/user/impl/RedisUser.kt
package dev.lbuddyboy.flash.user.impl

import dev.lbuddyboy.flash.handler.RedisHandler.Companion.requestJedis
import dev.lbuddyboy.flash.user.User
import dev.lbuddyboy.flash.user.model.Grant
import dev.lbuddyboy.flash.util.bukkit.*
import dev.lbuddyboy.flash.util.gson.GSONUtils
import java.util.*

class RedisUser(uuid: UUID?, name: String?, load: Boolean) : User() {
    init {
        this.uuid = uuid
        this.name = name
        if (load) load()
    }

    override fun load() {
        try {
            val user = GSONUtils.getGSON()
                .fromJson<RedisUser>(requestJedis().resource.hget("Users", getUuid().toString()), GSONUtils.REDIS_USER)
            name = user.getName()
            ip = user.getIp()
            permissions = user.getPermissions()
            knownIps = user.getKnownIps()
            punishments.addAll(user.getPunishments())
            notes = user.getNotes()
            grants = user.getGrants()
            promotions = user.getPromotions()
            demotions = user.getDemotions()
            activePrefix = user.getActivePrefix()
            playerInfo = user.getPlayerInfo()
            staffInfo = user.getStaffInfo()
            if (grants.isEmpty()) {
                grants.add(Grant.Companion.defaultGrant())
            }
            updateGrants()
        } catch (ignored: Exception) {
            save(true, true)
        }
    }

    override fun loadRank() {
        load()
    }

    override fun save(async: Boolean) {
        save(async, false)
    }

    private fun save(async: Boolean, reload: Boolean) {
        if (!async) {
            requestJedis().resource.hset("Users", getUuid().toString(), toJson())
            if (reload) load()
            return
        }
        Tasks.runAsync { save(false, reload) }
    }

    fun toJson(): String {
        return GSONUtils.getGSON().toJson(this, GSONUtils.REDIS_USER)
    }
}